Use log scale for visualizer

evanc577 opened the pull request:
This PR changes the visualizer to use log scales on both the frequency and magnitude axes. Fixes #326

The current visualizer uses linear scales for frequency and magnitude, but human perception of both is log scale. This means that most interesting things happen on the very left, while the bars in the center and right take up too much space in comparison to how we perceive them.

## Some notable changes in this PR:

- Log scale frequency and magnitude (required raising DFT size, currently 2^14)
- [Blackman window](https://en.wikipedia.org/wiki/Window_function#Blackman_window) for better sidelobe attenuation
- Adjustable visualizer parameters in visualizer.h
    - `const uint32_t DFT_SIZE = 1<<14;` (Total number of DFT bins)
    - `const uint32_t DFT_PAD = DFT_SIZE - 4000;` (Number of padded zeros, must be < DFT_SIZE)
    - `const double DYNAMIC_RANGE = 100;` (Dynamic range of visualization in dB)
    - `const double HZ_MIN = 20;` (Leftmost frequency of visualization)
    - `const double HZ_MAX = 20000;` (Rightmost frequency of visualization)
    - `const double GAIN = 0;` (Visualization gain in dB, tune if bars are too large/small)
- Added config option `visualizer_smooth_look` for smoother spectrum visualization
- Added config option `visualizer_cubic_interp` to use cubic interpolation between DFT bins when needed
- Fixed DFT calculation for stereo visualization
- Fixed DFT normalization
- Fixed reading from FIFO
- Removed some operations that I thought were arbitrary, such as buffing higher frequencies and cutting bandwidth

## Screenshot

### Default settings
Top: pulseeffects for reference
Bottom Left: ncmpcpp 0.8.2
Bottom Right: this PR
![ncmpcpp_visualizer](https://user-images.githubusercontent.com/14303970/87857206-ec47f100-c8f2-11ea-984c-61074607d995.png)

### `visualizer_smooth_look = yes`

![Screenshot_20200730_202952](https://user-images.githubusercontent.com/14303970/88987311-84ef5100-d2a3-11ea-85ed-f2d212e6ecea.png)

![visualizer_smooth](https://user-images.githubusercontent.com/14303970/88987960-8883d780-d2a5-11ea-9675-d03b5f025186.gif)

arybczak commented:
Thanks,  that does make the visualizer a lot better!

Can you add new configuration variables to example config file and describe them in the man page?

This is how the visualizer looks with smooth look, sound output that spans the majority of frequency range with pretty much maximum amplitude: ![new_visualizer](https://user-images.githubusercontent.com/387658/89187564-c3378980-d59d-11ea-8c66-9118e483858f.png)

(I'm using `visualizer_color = 47, 83, 119, 155, 191, 227, 221, 215, 209, 203, 197, 161` with 256 color terminal)

1. The smooth look is buggy for the bottom portion of the stereo visualiser, can this be fixed with appropriate symbols?
2. Looks like the bars never reach the top of the window, usually occupying about half. Can the visualiser be "stretched" a bit more?

evanc577 commented:
Yes, I can update the examples and man page. I think I'll also move the current hard-coded parameters into the configuration file to make tweaking the look of the visualizer easier.

To answer your questions:

1. I'm using [unicode block elements](https://en.wikipedia.org/wiki/Block_Elements) for smooth look. Unfortunately, the required characters for upper fractional blocks don't exist. I could add a config option for people with patched fonts where they could specify their own glyphs, but I think that's a very niche scenario, and that using stereo visualization along with smooth look should just be unsupported.
2. The `GAIN` and `DYNAMIC_RANGE` parameters can be adjusted. Currently I have set them to 0dB and 100dB respectively.
    * The current settings ensure that no clipping is possible. I'd guess that for most music, a +20dB gain could be used without clipping. Increasing/decreasing `GAIN` shifts the bars up/down. Increasing/decreasing `DYNAMIC_RANGE` widens/narrows the range of volumes displayed. To "strech" the visualizer, increase `GAIN` and decrease `DYNAMIC_RANGE` by the same amount.
    * My DSP is a bit rusty, but a couple factors that reduce the heights of the bars from maximum are:
        * non-pure tones (sinusoidals) can't reach 0dB
        * window function choice
        * scalloping loss when frequencies are between DFT bins
        * averaging across many bins for high frequencies.
    * One potential issue: larger bars inevitably requires more characters to be drawn. This seems to be a major performance bottleneck, especially when setting the character's color.

arybczak commented:
> Unfortunately, the required characters for upper fractional blocks don't exist

It looks like you can emulate that with NC::Format::Reverse. Transparency is lost for the rest of a block, but I think it's still better than not supporting the smooth look at all given how much better it is.

evanc577 commented:
Interesting hack with reverse video, IMO looks a bit weird with transparent terminals, but if users want it, then it's there for people that want it.

![Screenshot_20200807_175923](https://user-images.githubusercontent.com/14303970/89691926-3d3a7c00-d8d8-11ea-961c-b9d5c4806b7f.png)

arybczak commented on src/screens/visualizer.cpp:
```diff
@@ -509,21 +509,41 @@ void Visualizer::DrawFrequencySpectrum(int16_t *buf, ssize_t samples, size_t y_o
 		{
 			size_t y = flipped ? y_offset+j : y_offset+height-j-1;
-			auto c = toColor(j, height);
+			auto color = toColor(j, height);
```
You can use NC::FormattedColor here and avoid the `reverse` flag dance.

evanc577 replied:
Not exactly sure how FormattedColor works exactly. I ended up creating a new FormattedColor if I need reverse video because it doesn't look like I can easily add a format to an existing FormattedColor.

arybczak commented:
Thanks. I think it's good to be merged modulo one detail, namely the amount of new configuration variables.

Do we have to have so many?

It looks to me like `visualizer_smooth_look` (and `visualizer_hz_min`, `visualizer_hz_max`) are enough, The rest seem to:
- not have viable alternatives (`visualizer_smooth_chars`)
- have no significant influence on the visuals (`visualizer_cubic_interp`)
- be internal details and break visualization when fiddled with (`visualizer_dft_size`, `visualizer_dft_pad`)
- be too static to be useful (`visualizer_dynamic_range`, `visualizer_gain`)

The last ones when increased also just seem to result in a "fatter" base of the visualizer, so it's better to have visualization for half the window than just having it more uniformly filled with color. I guess that's a problem to be solved another day.

BTW, why did you disable autoscaling for spectrum? IMO autoscaling is much better than `visualizer_gain`, because it dynamically adjusts to the music.

evanc577 commented:
There doesn't *have* to be so many config variables, but I figured I would just expose them in case users want to mess with them.

> not have viable alternatives (visualizer_smooth_chars)

This is true, I can remove it if you think it's not needed.

> have no significant influence on the visuals (visualizer_cubic_interp)

With a large terminal window width and/or small dft size, some columns at lower frequencies don't have a corresponding DFT bin, thus we have to fake some data points there. The default `visualizer_cubic_interp = no` setting just reuses the previous column's height, which can look strange. `visualizer_cubic_interp = yes` is intended to smooth the columns in these cases, rather than trying to just brute force it with a higher DFT size.

I agree that just setting it on by default and removing the config option would work, but again, I figured I would just put it there because there wasn't any reason not to do so.

> be internal details and break visualization when fiddled with (visualizer_dft_size, visualizer_dft_pad)

Can you specify which values break visualization? I tried to do some basic checking before starting to catch bad values, but I may have missed some cases.

Some examples of changing these variables:

* `visualizer_dft_size = 16384`, `visualizer_dft_pad = 0`: causes the visualizer to look back farther in time, resulting in smoother/less jumpy bars
 * `visualizer_dft_size = [>16384]`: I have tested on a 2560x1440p monitor, and a 16384 DFT size seems to be acceptable. If people have wider monitors, they can set this to a higher value for more accuracy at low frequencies, or vice versa.

> be too static to be useful (visualizer_dynamic_range, visualizer_gain)

Figured I might as well expose them in case people want to use them.

---

As for the autoscaler, there are a few reasons I removed it.

1. Sometimes for songs that begin with a crescendo, the visualizer starts out very low, then suddenly jumps to a high value, which is jarring with the spectrum visualizer
2. One reason that a log scale magnitude spectrum visualizer is useful is that it is very easy to match a bar height with perceived loudness. Using an autoscaler makes this hard/impossible, especially for songs that have both quiet and loud sections. 
3. With my reworked fifo read algorithm, the current time-domain autoscaler is not technically correct. It's not noticeable in the waveform visualizers, but theoretically it may cause noticeable issues after a DFT. This could be worked around with a frequency-domain autoscaler.

I could toggle the spectrum autoscaler under a yet another config variable though...

evanc577 commented:
Any updates/thoughts on my last comment?

arybczak commented:
Eh, sorry.

> visualizer_dft_size = 16384, visualizer_dft_pad = 0: causes the visualizer to look back farther in time, resulting in smoother/less jumpy bars

If that's the case, can these two be merged and named `visualiser_dft_backlook` (or something similar, `visualiser_dft_smoothness` unfortunately seems to be too similar to `visualiser_smooth_look`) with a few integer values (perhaps 1,2,3) that internally translate to reasonable values of `dft_size` and `dft_pad`?

Other than that `visualizer_smooth_look`, `visualizer_hz_min` and `visualizer_hz_max` are alright. Please remove the rest because they will confuse people.

As for the autoscaler, would it work to add a config option to make it switchable in general (i.e. for all visualizers)? It can use the old method for everything or a different one for spectrum if you wish to add it. I just want consistent behavior for all types, i.e. either the autoscaler is on or not, not that it's silently disabled for spectrum.

If these are done, I'll merge as I really like this PR (thanks again for doing that :slightly_smiling_face: ).

evanc577 commented:
Removed/renamed config options, the remaining options I've added are:

```diff
diff --git a/src/settings.h b/src/settings.h
index 7738375..1890ac3 100644
--- a/src/settings.h
+++ b/src/settings.h
@@ -82,6 +82,11 @@ struct Configuration
        std::string lastfm_preferred_language;
        std::wstring progressbar;
        std::wstring visualizer_chars;
+       bool visualizer_autoscale;
+       bool visualizer_spectrum_smooth_look;
+       uint32_t visualizer_spectrum_dft_size;
+       double visualizer_spectrum_hz_min;
+       double visualizer_spectrum_hz_max;
 
        std::string pattern;
```

The `visualizer_spectrum_dft_size` config option now corresponds to an actual DFT size of 2^(visualizer_spectrum_dft_size) and is set to 14 by default (actual DFT size 16384).

The old `visualizer_dft_pad` option is removed. Instead it is computed as 2^(visualizer_spectrum_dft_size - 2).

The autoscaler is now globally toggle-able with `visualizer_autoscale`. Though with it enabled and a high dft_size, the spectrum visualizer appears kind of "jumpy", and due to the log scale magnitude of the visualizer, doesn't scale the output by an appreciable amount. It's fixable by applying a scaling algorithm in the frequency domain rather than time domain, but to do it correctly seems like a lot of additional complexity that I'm not willing to take on right now.

arybczak commented:
Thanks!

> The visualizer_spectrum_dft_size config option now corresponds to an actual DFT size of 2^(visualizer_spectrum_dft_size) and is set to 14 by default (actual DFT size 16384).

I checked it and it looks to me that usable values are in range [14, 19] (both 14 and 19 might be pushing it as 14 seems too jumpy, while 19 too laggy, but let's have them anyway). Can we reduce it to a configuration variable (let's call it `visualizer_spectrum_smoothness_factor`) that takes range [0, 5] that corresponds to [14, 19]? And let's make the default 16, i.e. 2.

Please also update config file :slightly_smiling_face:

evanc577 commented:
Not sure I like the `visualizer_spectrum_smoothness_factor` name. Seems to be a bit too similar to the existing `visualizer_spectrum_smooth_look`, and they do different things. For lack of better words, one is "temporal" smoothness while the other is "spacial" smoothness. My opinion is just to keep it as `visualizer_spectrum_dft_size` and put in a better description for what it does in the man/config files (thanks for the reminder for the config file).

Internally, I've changed some things around, mainly removing the zero padding for large dft sizes where they're not needed. This should mean less cpu usage for large dft sizes.

Unrelated, but what does the `visualizer_sync_interval` do? git blame leads me to commit e43ea18d29773c32bed4e3596ed2dfb4ec4a446d, but the link to http://musicpd.org/mantis/view.php?id=2503 is broken. Everything seems to work as intended if I comment that code block out. (I haven't commented it out in this PR, just wondering)

arybczak commented:
> My opinion is just to keep it as visualizer_spectrum_dft_size and put in a better description for what it does

That doesn't solve the problem that only [14, 19] seems to be the useful range. I'd like to have a setting where possible values make sense. Not to mention that for large values (eg 25) it crashes.

I'm fine with a different name of a setting, but please make it accept the range I mentioned above.

> Unrelated, but what does the visualizer_sync_interval do?

It's a dirty hack for keeping visualization synced with music (at least needed on Alsa output, on Pulse doesn't seem to be needed). It disables and enables the FIFO output every N seconds.

arybczak commented:
Oh, you modified the range. Sorry, I didn't notice. I don't have a better idea for a name, so if you also don't let's leave it.

I'll check it out tomorrow and merge if all is well 👍

arybczak commented on doc/config:
```diff
@@ -94,6 +94,31 @@
+#visualizer_spectrum_smooth_look = yes
+#
+## A value between 0 and 5 inclusive. Specifying a larger value makes the
+## visualizer look at a larger slice of time, which results less jumpy
```
```suggestion
## visualizer look at a larger slice of time, which results in less jumpy
```

arybczak commented on doc/ncmpcpp.1:
```diff
@@ -99,6 +99,21 @@ Defines visualizer's look (string has to be exactly 2 characters long: first one
+For spectrum visualizer, use unicode block characters for a smoother, more continuous look. This will override the visualizer_look option. With transparent terminals and visualizer_in_stereo set, artifacts may be visible on the bottom half of the visualization.
+.TP
+.B visualizer_spectrum_dft_size = NUMBER
+For spectrum visualizer, a value between 0 and 5 inclusive. Specifying a larger value makes the visualizer look at a larger slice of time, which results less jumpy visualizer output.
```
```suggestion
For spectrum visualizer, a value between 0 and 5 inclusive. Specifying a larger value makes the visualizer look at a larger slice of time, which results in less jumpy visualizer output.
```

arybczak commented on doc/config:
```diff
@@ -94,6 +94,31 @@
 #visualizer_color = 41, 83, 119, 155, 185, 215, 209, 203, 197, 161
 #
+##
+## Note: The next few visualization options apply to the spectrum visualizer
```
```suggestion
## Note: The next few visualization options apply to the spectrum visualizer.
```

arybczak commented on src/settings.cpp:
```diff
@@ -291,6 +291,26 @@ bool Configuration::read(const std::vector<std::string> &config_paths, bool igno
+			"12", [](std::string v) {
+      uint32_t result = verbose_lexical_cast<uint32_t>(v);
+      boundsCheck<uint32_t>(result, 0, 5);
+			return result + 12;
```
Shouldn't it be 14?

evanc577 replied:
This should be `result+12`

arybczak commented on src/settings.cpp:
```diff
@@ -291,6 +291,26 @@ bool Configuration::read(const std::vector<std::string> &config_paths, bool igno
+	p.add("visualizer_autoscale", &visualizer_autoscale, "yes", yes_no);
+	p.add("visualizer_spectrum_smooth_look", &visualizer_spectrum_smooth_look, "yes", yes_no);
+	p.add("visualizer_spectrum_dft_size", &visualizer_spectrum_dft_size,
+			"12", [](std::string v) {
```
```suggestion
			"2", [](std::string v) {
```

arybczak commented:
I'll merge and apply the suggestions I wrote.

Thanks a lot for this patch again :tada:

arybczak commented:
FWIW after experimentation I restored `visualizer_spectrum_gain` (and made it tied to dynamic_range) for the possibility to fill the window a bit more. See 261b232 for more details.

evanc577 commented:
Thanks for the merge! :)

One issue though, I see that you've helpfully fixed the default `visualizer_spectrum_dft_size` for me in 28382a09b9f500c79f8497a4516d084353353677. But the correct return value should be `return result + 12;`, not `return result + 14;` I had made some last minute internal changes, the new 12 is effectively the same as the old 14.

arybczak commented:
Oh, I see. Thanks, I widened the range in ee66477 and after a bit more pondering set 15 to be the default as it's a bit more dynamic.
